Civi/Api4/Action/Create.php

```diff
@@ -40,6 +40,13 @@
  */
 class Create extends AbstractAction {
 
+  /**
+   * And one of the field in the set is required
+   *
+   * @var array
+   */
+  protected $optionalRequiredFields = [];
```

Comment:
@colemanw this special variable to support ```civicrm_api3_verify_one_mandatory``` where either one the attribute is mandatory. Do you agree with this is the approach?

Reply:
I think adding it as a protected variable makes it an api param and exposes it to the api explorer. We don't want that.

I would prefer a metadata-driven approach. @monishdeb @eileenmcnaughton what would you think about adding a php-like syntax to a specal `requiredIf` fieldspec property, like this:

    $fields = array(
      'contact_type' => [
        'title' => 'Contact Type',
        'data_type' => 'String',
        'required' => true,
      ],
      'first_name' => [
        'title' => 'First Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Individual' && !$last_name && !$email",
      ],
      'last_name' => [
        'title' => 'Last Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Individual' && !$first_name && !$email",
      ],
      'organization_name' => [
        'title' => 'Organization Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Organization' && !$email",
      ],
      'household_name' => [
        'title' => 'Household Name',
        'data_type' => 'String',
        'requiredIf' => "$contact_type == 'Household' && !$email",
      ],
    );

Reply:
We might even be able to use the Smarty engine to parse those conditional strings.

Reply:
I like this approach, and if I am not wrong it should extend the ``validateValues``` and use ```eval()``` to translate the expression.

Reply:
I think the evaluator code could look something like this:

    function evaluateExpression($expr, $vars) {
      var $smarty = new Smarty();
      foreach ($vars as $key => $val) {
        $smarty->assign($key, $val);
      }
      $tpl = "{if $expr}1{else}0{/if}";
      $result = trim($smarty->fetch('eval:' . $tpl));
      return $result;
    }